fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,336 @@
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(NO_INTELLISENSE) && !defined(FDBSERVER_TENANTENTRYCACHE_ACTOR_G_H)
+#define FDBSERVER_TENANTENTRYCACHE_ACTOR_G_H
+#include "fdbserver/TenantEntryCache.actor.g.h"
+#elif !defined(FDBSERVER_TENANTENTRYCACHE_ACTOR_H)
+#define FDBSERVER_TENANTENTRYCACHE_ACTOR_H
+
+#pragma once
+
+#include "fdbclient/DatabaseContext.h"
+#include "fdbclient/FDBOptions.g.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/RunTransaction.actor.h"
+#include "fdbclient/Tenant.h"
+#include "fdbclient/TenantManagement.actor.h"
+#include "fdbserver/Knobs.h"
+#include "flow/IndexedSet.h"
+
+#include <functional>
+#include <unordered_map>
+
+#include "flow/actorcompiler.h" // has to be last include
+
+#define INITIAL_CACHE_GEN 1
+
+using TenantEntryCacheGen = uint64_t;
+using TenantNameEntryPair = std::pair<TenantName, TenantMapEntry>;
+using TenantNameEntryPairVec = std::vector<TenantNameEntryPair>;
+
+enum class TenantEntryCacheRefreshReason { INIT = 1, PERIODIC_TASK = 2, CACHE_MISS = 3 };
+
+template <class T>
+struct TenantEntryCachePayload {
+	TenantName name;
+	TenantMapEntry entry;
+	TenantEntryCacheGen gen;
+	// Custom client payload
+	T payload;
+};
+
+template <class T>
+using TenantEntryCachePayloadFunc =
+    std::function<TenantEntryCachePayload<T>(const TenantName&, const TenantMapEntry&, const TenantEntryCacheGen)>;
+
+// In-memory cache for TenantEntryMap objects. It supports three indices:
+// 1. Lookup by 'TenantId'
+// 2. Lookup by 'TenantPrefix'
+// 3. Lookup by 'TenantName'
+//
+// TODO:
+// ----
+// The cache allows user to construct the 'cached object' by supplying a callback to allow constructing the cached
+// object. The cache implements a periodic refresh mechanism, polling underlying database for updates (add/remove
+// tenants), in future we might want to implement database range-watch to monitor such updates
+
+template <class T>
+class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyable {
+private:
+	UID uid;
+	Database db;
+	TenantEntryCacheGen curGen;
+	TenantEntryCachePayloadFunc<T> createPayloadFunc;
+
+	AsyncTrigger sweepTrigger;
+	Future<Void> sweeper;
+	Future<Void> refresher;
+	Map<int64_t, TenantEntryCachePayload<T>> mapByTenantId;
+	Map<TenantName, TenantEntryCachePayload<T>> mapByTenantName;
+
+	// TODO: add hit/miss/refresh counters
```

Comment:
miss by reason would be nice too since you already have those reasons enumerated